tree2: add TypedTreeFactory

CraigMacomber opened the pull request:
## Description

Adds wrapper exposing minimal but fully schema aware API for shared tree.

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

CraigMacomber committed:
tree2: add TypedTreeFactory

noencke commented on experimental/dds/tree2/src/typed-tree/typedTree.ts:
```diff
@@ -0,0 +1,131 @@
+ * IChannel wrapper that exposes a "root".
+ *
+ * @remarks
+ * This is handy when an implementing IChannelFactory, and want to return a type thats derived from another IChannel implementation.
```
```suggestion
 * This is handy when an implementing IChannelFactory and it's desirable to return a type that's derived from another IChannel implementation.
```

anthony-murphy commented:
The pattern here is a bit odd. generally, our factories allow building any dds of that type, so for example, a map dds builds any map, so i only have one factory entry, regardless of the number of maps that i have.  this seems to bind a factory to a schema, so i can't build any tree from this factory, only trees with a specific schema. 

did you consider late binding the schema? so i the factory creates any tree, even if that tree has a defined internal schema, but we limit usage of that tree, at least via api, until the schema is bound, which would be after load/create.

its a bit of a different pattern, as i think you only expect one tree, but i'm also not sure our consumer will only have one tree. In fact with this mechanism it doesn't seem possible to have multiple trees, as the type for factory is static, so it won't work to have multiple factories for multiple tree with different schema.

anthony-murphy commented:
another note. i've been thinking about azure/odsp client, and i think it will be even harder to make a pattern like this work, as it make assumptions about how dds expose factories. specifically, it assumes it will be static

CraigMacomber committed:
Update experimental/dds/tree2/src/typed-tree/typedTree.ts

Co-authored-by: Noah Encke <78610362+noencke@users.noreply.github.com>

CraigMacomber committed:
Merge branch 'main' of https://github.com/microsoft/FluidFramework into TypedTree

CraigMacomber committed:
Add subtype to TypedTreeOptions

CraigMacomber commented:
> The pattern here is a bit odd. generally, our factories allow building any dds of that type, so for example, a map dds builds any map, so i only have one factory entry, regardless of the number of maps that i have. this seems to bind a factory to a schema, so i can't build any tree from this factory, only trees with a specific schema.

The idea is apps can define effectively define customized tree types. Though as you point out, this wasn't properly implemented for the azure/factory registry use case since they all got the same name. I have fixed this now by adding a "subtype". 

> did you consider late binding the schema? so the factory creates any tree, even if that tree has a defined internal schema, but we limit usage of that tree, at least via api, until the schema is bound, which would be after load/create.

This approach has been considered, and is still an option we way want to revisit. I have no added TODOs indicating we may want to take that approach instead. Personally, I like lumping that DDS data format compatibility checks in with the app's compatibility checks (via schema) since it provides a single place for compatibility errors, but once shared tree actually has a proper error model for how we handle compatibility errors, it will make sense to reevaluate this.

> its a bit of a different pattern, as i think you only expect one tree, but i'm also not sure our consumer will only have one tree. In fact with this mechanism it doesn't seem possible to have multiple trees, as the type for factory is static, so it won't work to have multiple factories for multiple tree with different schema.

I have now fixed the fact that "type" is hard coded. Good catch.

CraigMacomber commented:
> another note. i've been thinking about azure/odsp client, and i think it will be even harder to make a pattern like this work, as it make assumptions about how dds expose factories. specifically, it assumes it will be static

The app can expose the actual factories they want (statically) as if they were creating specialized DDSs. We already had this requirement for the already existing tree and tree2 factories due to them containing configuration for what write format to use for compatibility control, so it isn't new in this change: there is just a bit more configuration included now.
